Add package ob-p5js

alejandrogallo opened the pull request:
### Brief summary of what the package does

This package provides a minor mode for p5js
and a way to export javascript code to an iframe
containing a p5js ready environment to export to html.

### Direct link to the package repository

https://github.com/alejandrogallo/ob-p5js
https://alejandrogallo.github.io/blog/posts/ob-p5js/

### Your association with the package

Maintainer and author.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
  ```
  4 issues found:

  1:0: warning: The package summary should start with an uppercase letter or a digit.
  57:0: error: "org-babel-header-args:p5js" doesn't start with package's prefix "ob-p5js".
  57:0: error: `org-babel-header-args:p5js' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
  187:0: error: "p5js-mode" doesn't start with package's prefix "ob-p5js".
  ```
- [X] My elisp byte-compiles cleanly
- [X] `M-x checkdoc` is happy with my docstrings
  - I am using literate programming, it does not pass the test but it would
    pass it were it not for some literate programming quirks.
- [X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this.  It's not a good idea to define a major mode called `p5js-mode` inside a package called `ob-p5js`; someone else might come along and create a `p5js-mode`, for example, and they'd run into conflicts.  Otherwise, here's a quick first pass --

`ob-p5js.el` with byte-compile using Emacs 28.1:
```
ob-p5js.el:36:1: Warning: defcustom for `org-babel-default-header-args:p5js'
    fails to specify type
ob-p5js.el:36:1: Warning: defcustom for `org-babel-default-header-args:p5js'
    fails to specify containing group
ob-p5js.el:36:1: Warning: defcustom for `org-babel-default-header-args:p5js'
    fails to specify type
ob-p5js.el:36:1: Warning: defcustom for `org-babel-default-header-args:p5js'
    fails to specify containing group
ob-p5js.el:57:1: Warning: defcustom for `org-babel-header-args:p5js' fails to
    specify type
ob-p5js.el:57:1: Warning: defcustom for `org-babel-header-args:p5js' fails to
    specify containing group
ob-p5js.el:57:1: Warning: defcustom for `org-babel-header-args:p5js' fails to
    specify type
ob-p5js.el:57:1: Warning: defcustom for `org-babel-header-args:p5js' fails to
    specify containing group
ob-p5js.el:72:1: Warning: defcustom for `ob-p5js-src' fails to specify
    containing group
ob-p5js.el:72:1: Warning: defcustom for `ob-p5js-src' fails to specify
    containing group
ob-p5js.el:84:1: Warning: defcustom for `ob-p5js-iframe-class' fails to
    specify type
ob-p5js.el:84:1: Warning: defcustom for `ob-p5js-iframe-class' fails to
    specify containing group
ob-p5js.el:84:1: Warning: defcustom for `ob-p5js-iframe-class' fails to
    specify type
ob-p5js.el:84:1: Warning: defcustom for `ob-p5js-iframe-class' fails to
    specify containing group
```

`ob-p5js.el` with checkdoc 0.6.2:
```
ob-p5js.el:0: The first line should be of the form: ";;; package --- Summary"
```

`ob-p5js.el` with package-lint 20221003.1636:
```
4 issues found:
1:0: warning: The package summary should start with an uppercase letter or a digit.
57:0: error: "org-babel-header-args:p5js" doesn't start with package's prefix "ob-p5js".
57:0: error: `org-babel-header-args:p5js' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
187:0: error: "p5js-mode" doesn't start with package's prefix "ob-p5js".
```

`ob-p5js.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- ob-p5js.el#L10: Have you done the paperwork to assign this copyright?
```

Package and license:
- GNU General Public License v3.0 (via API)
- unknown license -- ob-p5js.el: (no header)
- Add *formal* license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to ob-p5js.el

riscy commented:
Also:

> I am using literate programming, it does not pass the test but it would pass it were it not for some literate programming quirks.

Avoid using literate programming unless you can get the checks to pass with the resulting source code.

riscy commented:
I'll mark this awaiting-upstream for now, but feel free to reach out.

alejandrogallo commented:
Sorry, I did not find time to work on this.

- So right now the `package-lint-current-buffer` says

  ```
  2 issues found:

  51:0: error: "org-babel-header-args:p5js" doesn't start with package's prefix "ob-p5js".
  51:0: error: `org-babel-header-args:p5js' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
  ```

  which I don't know how to fix since these variables are a must for such a package.

- checkdoc has no errors due to literate programming now.

- I deleted the `p5js-mode` part which I only used so that the src blocks
  have a javascript environment.
  I replaced it with an `ob-p5js-mode` which should be ok also maybe the user
  wants to define some bindings for these source blocks.
  

I think from my side then this should be done, right?
This would be the final version then
https://github.com/alejandrogallo/ob-p5js/blob/main/ob-p5js.el

Thank you for your time and commitment

riscy commented:
You're right - I'm constantly re-reminding myself about org-babel's conventions.  :)  Disregard those prefix lints for this package.

Here's the latest --

`ob-p5js.el` with byte-compile using Emacs 28.1:
```
ob-p5js.el:29:1: Warning: defcustom for `org-babel-default-header-args:p5js'
    fails to specify type
ob-p5js.el:29:1: Warning: defcustom for `org-babel-default-header-args:p5js'
    fails to specify type
ob-p5js.el:51:1: Warning: defcustom for `org-babel-header-args:p5js' fails to
    specify type
ob-p5js.el:51:1: Warning: defcustom for `org-babel-header-args:p5js' fails to
    specify type
ob-p5js.el:86:1: Warning: defcustom for `ob-p5js-src' fails to specify
    containing group
ob-p5js.el:86:1: Warning: defcustom for `ob-p5js-src' fails to specify
    containing group
ob-p5js.el:98:1: Warning: defcustom for `ob-p5js-iframe-class' fails to
    specify type
ob-p5js.el:98:1: Warning: defcustom for `ob-p5js-iframe-class' fails to
    specify containing group
ob-p5js.el:98:1: Warning: defcustom for `ob-p5js-iframe-class' fails to
    specify type
ob-p5js.el:98:1: Warning: defcustom for `ob-p5js-iframe-class' fails to
    specify containing group
```
To get rid of the "fails to specify containing group" nags, you can use something like
```elisp
(defgroup ob-p5js nil
  "Some-description-here."
  :prefix "ob-p5js-"
  :group 'org-babel)
```
near the top of the file, which will automatically assign the subsequent defcustoms in the rest of the file to that group (you may want to update the existing `:group` keywords to point to the `ob-p5js` group too).


`ob-p5js.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- ob-p5js.el#L3: Have you done the paperwork to assign this copyright?
```
I ask this one because sometimes people copy-paste this copyright notice without actually having assigned that copyright over.

Package and license:
- Add *formal* license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to ob-p5js.el

riscy commented:
I'll mark this awaiting upstream for now, but feel free to reach out!

alejandrogallo commented:
Thank you for your time, I fixed the warnings, the only thing I don't know quite to do is about the paperwork
for the copyright?

```
- ob-p5js.el#L3: Have you done the paperwork to assign this copyright?
```
maybe someone can give me a couple of pointers?

riscy commented:
Apologies for leaving this unanswered!  I hope this hasn't been too much inconvenience.

> maybe someone can give me a couple of pointers?

If you haven't done any paperwork, `Copyright (C) 2022 Free Software Foundation, Inc.` is most likely not true; the copyright is your own, not the FSF's, until you explicitly sign it to them.

Since the last pass it looks like a lot has been fixed.  Here's the latest --

`ob-p5js.el` with byte-compile using Emacs 28.1:
```
ob-p5js.el:42:1: Warning: defcustom for `org-babel-default-header-args:p5js'
    fails to specify type
ob-p5js.el:42:1: Warning: defcustom for `org-babel-default-header-args:p5js'
    fails to specify type
ob-p5js.el:64:1: Warning: defcustom for `org-babel-header-args:p5js' fails to
    specify type
ob-p5js.el:64:1: Warning: defcustom for `org-babel-header-args:p5js' fails to
    specify type
```

`ob-p5js.el` with package-lint 20221003.1636:
```
2 issues found:
64:0: error: "org-babel-header-args:p5js" doesn't start with package's prefix "ob-p5js".
64:0: error: `org-babel-header-args:p5js' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
```

- Add *formal* license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to ob-p5js.el

alejandrogallo commented:
Thank you for the info. Today I had some time on my hands so I addressed the byte compile issues, I replaced the defcustoms by defvar and defconst as I've seen other ob packages do.

I changed the license to MIT. I hope now this should be good to go?

riscy commented:
Looks good -- rather than:
```
;; Licence: MIT
```
I'd recommend
```
;; SPDX-License-Identifier: MIT
```
which is machine-detectable and unambiguous.  But since it's been a while I trust you will get to that.

Thanks again for this -- and welcome to MELPA. :)

alejandrogallo commented:
Oh thanks, I just changed the LICENSE part in https://github.com/alejandrogallo/ob-p5js/commit/c9f89c8b588cdb0c021a875b8f9fc91be7e39de8

Thank you for your patience, I hope the next packages coming from me represent less work for you, I really appreciate your effort and time.

Cheers
